core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -232,6 +232,9 @@ public Snapshot apply() {
 
     ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
 
+    // merge flinkCheckpointInfo from previousSummary into currentSummary when rewrite datafiles
+    mergeFlinkCheckpointInfo(builder, previousSummary, summary);
```

Comment:
I don't think that we want to modify core to fix this. Instead of carrying metadata forward, I think we should instead look at older snapshots to find the max committed ID.